lib/elixir/test/elixir/typespec_test.exs

```diff
@@ -1183,6 +1183,53 @@ defmodule TypespecTest do
       end)
     end
 
+    test "spec_to_quoted with maps with __struct__ key" do
+      defmodule A do
+        defstruct [:key]
+      end
+
+      defmodule B do
+        defstruct [:key]
+      end
+
+      bytecode =
+        test_module do
+          @spec single_struct(%A{}) :: {:ok}
+          def single_struct(arg), do: {:ok, arg}
+
+          @spec single_struct_key(%{__struct__: A}) :: {:ok}
+          def single_struct_key(arg), do: {:ok, arg}
+
+          @spec union_struct(%A{} | %B{}) :: {:ok}
+          def union_struct(arg), do: {:ok, arg}
+
+          @spec union_struct_key(%{__struct__: A | B}) :: {:ok}
+          def union_struct_key(arg), do: {:ok, arg}
```

Comment:
This is the only one that actually needs to be tested. But I wanted to make sure the other ones will pass so I wrote the test for the. Let me know if we don't want them all, and I will remove them
/cc @ericmj